spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/StreamDefinition.java

```diff
@@ -153,6 +149,28 @@ public void remove() {
 			throw new UnsupportedOperationException();
 		}
 
+		/**
+		 * Return whether or not this iterator is past the very first module to deploy.
+		 */
+		public boolean hasMoreDownstream() {
+			return i < modules.size() - 2;
+		}
+
+		/**
+		 * Return whether or not there are more modules to deploy.
+		 */
+		public boolean hasMoreUpstream() {
+			return hasNext();
+		}
+
+		/**
+		 * Return the next module to deploy after the one currently pointed at, if any.
+		 * This does not advance this iterator.
+		 */
+		public ModuleDefinition peekUpstream() {
+			return i >= 0 ? modules.get(i) : null;
```

Comment:
nit: could call `hasMoreUpstream()` instead of `i >= 0` here (possibly this could evolve, and the `hasMoreUpstream()` method would always be accurate)